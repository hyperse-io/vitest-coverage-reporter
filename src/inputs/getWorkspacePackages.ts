import { getPackages } from '@manypkg/get-packages';

/**
 * Get the workspace packages
 * @param cwd The current working directory, repo root
 * @returns
 */
export async function getWorkspacePackages(
  cwd: string
): Promise<
  Map<string, { name: string; relativeDir: string; version: string }>
> {
  const { packages } = await getPackages(cwd);
  const sortedPackages: Array<
    [string, { name: string; relativeDir: string; version: string }]
  > = packages.map((x) => [
    x.dir,
    {
      name: x.packageJson.name,
      version: x.packageJson.version,
      relativeDir: x.relativeDir,
    },
  ]);
  // sort by package name
  sortedPackages.sort((a, b) => {
    if (a[1].name < b[1].name) {
      return -1;
    }
    if (a[1].name > b[1].name) {
      return 1;
    }
    return 0;
  });

  return new Map(sortedPackages);
}
