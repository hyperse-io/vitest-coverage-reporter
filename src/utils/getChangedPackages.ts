import { getPackages, Package } from '@manypkg/get-packages';

export async function getChangedPackages(
  cwd: string,
  workspacePackages: Map<string, { name: string; version: string }>
) {
  const { packages } = await getPackages(cwd);
  const changedPackages = new Set<Package>();

  for (const pkg of packages) {
    const previousPackages = workspacePackages.get(pkg.dir);
    if (previousPackages?.version !== pkg.packageJson.version) {
      changedPackages.add(pkg);
    }
  }

  return [...changedPackages];
}
