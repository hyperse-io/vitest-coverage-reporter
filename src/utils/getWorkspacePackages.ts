import { getPackages } from '@manypkg/get-packages';

export async function getWorkspacePackages(cwd: string) {
  const { packages } = await getPackages(cwd);
  const sortedPackages: Array<[string, { name: string; version: string }]> =
    packages.map((x) => [
      x.dir,
      {
        name: x.packageJson.name,
        version: x.packageJson.version,
      },
    ]);
  // sort by package name
  sortedPackages.sort((a, b) => {
    if (a[1].name < b[1].name) {
      return -1;
    }
    if (a[1].name > b[1].name) {
      return 1;
    }
    return 0;
  });

  return new Map(sortedPackages);
}
