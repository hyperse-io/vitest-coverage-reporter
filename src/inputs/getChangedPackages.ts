import { Package } from '@manypkg/get-packages';
import { FileCoverageMode } from './getCoverageModeFrom';
import { getPullChanges } from './getPullChanges';
import { getWorkspacePackages } from './getWorkspacePackages';

export async function getChangedPackages(repoCwd: string) {
  const workspacePackages = await getWorkspacePackages(repoCwd);
  const changedPackages = new Set<Package>();
  const allChangedFiles = await getPullChanges(FileCoverageMode.All);

  for (const [dir, { name, relativeDir, version }] of workspacePackages) {
    const packageChanged = allChangedFiles.find(
      (s) => !!~s.indexOf(relativeDir)
    );

    if (packageChanged) {
      changedPackages.add({
        dir,
        relativeDir,
        packageJson: {
          name,
          version,
        },
      });
    }
  }

  return [...changedPackages];
}
