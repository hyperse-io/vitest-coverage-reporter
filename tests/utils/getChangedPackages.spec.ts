import { join } from 'node:path';
import { getChangedPackages } from '../../src/utils/getChangedPackages.js';
import { getWorkspacePackages } from '../../src/utils/getWorkspacePackages.js';

describe('getChangedPackages()', () => {
  const fixtureCwd = join(__dirname, 'fixtures');
  it('Should correct resolve file changed packages', async () => {
    const allWorkspacePackages = await getWorkspacePackages(fixtureCwd);
    const changedPackages = await getChangedPackages(
      fixtureCwd,
      allWorkspacePackages
    );
    expect(changedPackages).toMatchSnapshot();
  });
});
