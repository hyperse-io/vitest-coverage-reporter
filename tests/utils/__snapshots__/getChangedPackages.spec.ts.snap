// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getChangedPackages() > Should correct resolve file changed packages 1`] = `[]`;
